### PR TITLE
ARROW-11944: [Developer] Fix archery's comparison of cached benchmark runs

### PR DESCRIPTION
**Pull request by bkietz**
Also added a unit test and some notes to `benchmarks.rst`.

My own preferred benchmarking workflow is to exclusively output run results to json and run `archery benchmark diff` against those. I'm not sure that automatic (re)generation and rebuild of the cmake build directory is a good default; IMO a new user will most likely be trying to iterate rather than make a single shot run and should be protected from staring at a release build log after every change. Is there enough guidance in `benchmarks.rst`?

@dianaclarke @edponce

### CODE REVIEW

File: dev/archery/archery/benchmark/runner.py [hunk 1]
```diff
@@ -120,11 +120,14 @@ def is_json_result(cls, path_or_str):
 
     @staticmethod
     def from_json(path_or_str, **kwargs):
-        # breaks recursive imports
-        from ..utils.codec import BenchmarkRunnerCodec
-        path_or_str, json_load = (open(path_or_str), json.load) \
-            if os.path.isfile(path_or_str) else (path_or_str, json.loads)
-        return BenchmarkRunnerCodec.decode(json_load(path_or_str), **kwargs)
+        # .codec imported here to break recursive imports
+        from .codec import BenchmarkRunnerCodec
```
Comment (dianaclarke) [inline dev/archery/archery/benchmark/runner.py:124]: Oh! This is my bad. Thanks so much for fixing it!

File: docs/source/developers/benchmarks.rst [hunk 2]
```diff
@@ -45,6 +45,11 @@ Sometimes, it is required to pass custom CMake flags, e.g.
   export CC=clang-8 CXX=clang++8
   archery benchmark run --cmake-extras="-DARROW_SIMD_LEVEL=NONE"
 
+Additionally a full CMake build directory may be specified.
+
+.. code-block:: shell
+  archery benchmark run $HOME/arrow/cpp/release-build
```
Comment (pitrou) [inline docs/source/developers/benchmarks.rst:51]: Thanks, I had never noticed this feature.

**Review by pitrou — APPROVED** [review]

File: docs/source/developers/benchmarks.rst [hunk 3]
```diff
@@ -103,6 +123,16 @@ expressions.
...
+Instead of rerunning benchmarks on comparison, a JSON file (generated by
+``archery benchmark run``) may be specified for the contender and/or the
+baseline.
+
+.. code-block:: shell
+  archery benchmark run --output=baseline.json $HOME/arrow/cpp/release-build
+  git checkout some-feature
+  archery benchmark run --output=contender.json $HOME/arrow/cpp/release-build
+  archery benchmark diff contender.json baseline.json
+
```
Comment (edponce) [inline docs/source/developers/benchmarks.rst:135]: Thanks for adding this info. I will most likely benefit from the capability of saving benchmark results in JSON files and being able to compare them at will.

File: dev/archery/archery/benchmark/runner.py (hunk 1 above)
Comment (edponce) [reply dev/archery/archery/benchmark/runner.py:124]: I do not understand the "break recursive imports" comment. I do not see how this import breaks the function from being called unless it has side-effects which I am not aware of.

**Comment by bkietz** [comment]
[CI failure](https://github.com/apache/arrow/pull/9712/checks?check_run_id=2114719851#step:8:9107) is unrelated ( https://issues.apache.org/jira/browse/ARROW-11717 ).

File: dev/archery/archery/benchmark/runner.py (hunk 1 above)
Comment (pitrou) [reply dev/archery/archery/benchmark/runner.py:124]: In Python, recursive imports are only bothersome at the top level. Otherwise they work just fine.

**Pull request closed without merging**